File: fleetmind/runtime/provisioner.py
"""
FleetMind provisioner — creates agent workspaces and installs skills.
"""

from __future__ import annotations
import shutil
from pathlib import Path

from rich.console import Console
from rich.progress import Progress, SpinnerColumn, TextColumn

from ..coordinator.config import FleetConfig, AgentConfig, SkillRef

console = Console()

AGENTS_MD_TEMPLATE = """\
# AGENTS.md — {name} ({emoji})

{description}

## Role

{"Orchestrator: coordinates the fleet and delegates to specialist agents." if orchestrator else "Specialist: handles delegated tasks from the orchestrator."}

## Memory

- Daily notes: `memory/YYYY-MM-DD.md`
- Long-term: `MEMORY.md`

## Tools

Skills define available tools. Check TOOLS.md for environment-specific notes.
"""

SOUL_MD_TEMPLATE = """\
# SOUL.md — {name}

{soul}
"""

USER_MD_TEMPLATE = """\
# USER.md — About Your Human

_Populated during onboarding._

- **Name:**
- **Timezone:**
- **Notes:**
"""


def provision_fleet(fleet: FleetConfig, dry_run: bool = False) -> None:
    """Provision all agent workspaces in the fleet."""
    console.print(f"\n[bold cyan]FleetMind[/bold cyan] — provisioning fleet [bold]{fleet.name}[/bold]")
    console.print(f"  Client: {fleet.client}")
    console.print(f"  Agents: {len(fleet.agents)}")
    console.print(f"  Dry run: {dry_run}\n")

    with Progress(
        SpinnerColumn(),
        TextColumn("[progress.description]{task.description}"),
        console=console,
    ) as progress:
        for agent in fleet.agents:
            task = progress.add_task(f"Provisioning {agent.emoji} {agent.name}...", total=None)
            _provision_agent(fleet, agent, dry_run=dry_run)
            progress.update(task, description=f"[green]✓[/green] {agent.emoji} {agent.name}")
            progress.stop_task(task)

    console.print("\n[green]✓ Fleet provisioned.[/green]")
    console.print("  Next: run [bold]fleetmind render[/bold] to generate openclaw.json")


def _provision_agent(fleet: FleetConfig, agent: AgentConfig, dry_run: bool = False) -> None:
    """Provision a single agent's workspace."""
    workspace = Path(fleet.defaults.workspace_base) / f"workspace-{agent.id}"

    if not dry_run:
        workspace.mkdir(parents=True, exist_ok=True)

    # Write SOUL.md
    soul_content = SOUL_MD_TEMPLATE.format(name=agent.name, soul=agent.persona.soul)
    _write_file(workspace / "SOUL.md", soul_content, dry_run)

    # Write AGENTS.md
    agents_content = AGENTS_MD_TEMPLATE.format(
        name=agent.name,
        emoji=agent.emoji,
        description=agent.description,
        orchestrator=agent.orchestrator,
    )
    _write_file(workspace / "AGENTS.md", agents_content, dry_run)

    # Write USER.md (only if it doesn't exist — don't overwrite customized versions)
    user_md = workspace / "USER.md"
    if not user_md.exists():
        _write_file(user_md, USER_MD_TEMPLATE, dry_run)

    # Write IDENTITY.md
    role = "Orchestrator" if agent.orchestrator else "Specialist"
    identity_content = f"""# IDENTITY.md

- **Name:** {agent.name}
- **Emoji:** {agent.emoji}
- **Role:** {role}
- **Description:** {agent.description}
"""
    _write_file(workspace / "IDENTITY.md", identity_content, dry_run)

    # Install skills
    if agent.skills:
        _install_skills(fleet, agent, workspace, dry_run)


def _install_skills(
    fleet: FleetConfig,
    agent: AgentConfig,
    workspace: Path,
    dry_run: bool,
) -> None:
    """Install skills into an agent workspace."""
    skills_dir = workspace / "skills"
    if not dry_run:
        skills_dir.mkdir(exist_ok=True)

    for skill_ref in agent.skills:
        _install_skill(fleet, skill_ref, skills_dir, dry_run)


def _install_skill(
    fleet: FleetConfig,
    skill_ref: SkillRef,
    skills_dir: Path,
    dry_run: bool,
) -> None:
    """Install a single skill. Tries local first, then skills repo."""
    skill_name = skill_ref.name

    # Try local skills dir first
    local_candidates = [
        Path("./skills") / skill_name,
        Path(fleet.skills_repo.local or "") / skill_name if fleet.skills_repo.local else None,
    ]

    for candidate in local_candidates:
        if candidate and candidate.exists():
            if not dry_run:
                dest = skills_dir / skill_name
                if dest.exists():
                    shutil.rmtree(dest)
                shutil.copytree(candidate, dest)
            return

    # Skills repo cloning (stub)
    if fleet.skills_repo.url:
        console.print(
            f"    [yellow]→[/yellow] Skill [bold]{skill_name}[/bold]: "
            f"remote install from {fleet.skills_repo.url} (coming soon)"
        )
        return

    console.print(f"    [yellow]⚠[/yellow] Skill [bold]{skill_name}[/bold] not found locally — skipping")


def diff_fleet(fleet: FleetConfig) -> list[str]:
    """Compare what fleet.yaml would deploy vs what's currently on disk. Returns change descriptions."""
    changes = []

    for agent in fleet.agents:
        workspace = Path(fleet.defaults.workspace_base) / f"workspace-{agent.id}"

        if not workspace.exists():
            changes.append(f"[+] Create workspace for {agent.emoji} {agent.name} at {workspace}")
            continue

        soul_path = workspace / "SOUL.md"
        if not soul_path.exists():
            changes.append(f"[+] {agent.name}: create SOUL.md")
        else:
            current = soul_path.read_text()
            expected = SOUL_MD_TEMPLATE.format(name=agent.name, soul=agent.persona.soul)
            if current.strip() != expected.strip():
                changes.append(f"[~] {agent.name}: update SOUL.md")

        skills_dir = workspace / "skills"
        for skill_ref in agent.skills:
            skill_path = skills_dir / skill_ref.name
            if not skill_path.exists():
                ver = f"@{skill_ref.version}" if skill_ref.version else "@latest"
                changes.append(f"[+] {agent.name}: install skill {skill_ref.name}{ver}")

    if not changes:
        changes.append("No changes detected.")

    return changes


def _write_file(path: Path, content: str, dry_run: bool) -> None:
    """Write a file, creating parent dirs. Skips if dry_run."""
    if dry_run:
        return
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(content)

File: src/runtime/provisioner.ts
/**
 * FleetMind provisioner — creates agent workspaces and installs skills.
 *
 * Skills are resolved via the three-tier resolver:
 *   clawhub → public ClaWHub/GitHub skill
 *   private → Continuous Agentics private registry (GitHub Packages)
 *   client  → client's own skills_repo (default)
 */

import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import type { Fleet, AgentConfig } from "../config/schema.js";
import { installSkill } from "./resolver.js";
import { slackChannel } from "../core/channels.js";
import { log } from "../utils/log.js";
import { workspaceTemplatePath } from "./bot-types.js";

// =============================================================================
// Role-template resolution
// =============================================================================

// =============================================================================
// Fleet roster helpers
// =============================================================================

const ROLE_LABELS: Record<string, string> = {
  "pm": "PM (orchestrator)",
  "backend-worker": "Backend worker",
  "frontend-worker": "Frontend worker",
  "worker": "Worker",
};

function roleLabel(role: string | undefined): string {
  return ROLE_LABELS[role ?? "worker"] ?? "Worker";
}

/**
 * Generate the `## Fleet Members` markdown section for a given agent.
 * Lists all other agents in the fleet (not self).
 */
export function buildFleetRoster(fleet: Fleet, currentAgent: AgentConfig): string {
  const peers = fleet.agents.list.filter((a) => a.id !== currentAgent.id);

  if (peers.length === 0) {
    return [
      "## Fleet Members",
      "",
      `You are a solo bot in the **${fleet.fleet.name}** fleet. No peer bots configured today.`,
      "",
    ].join("\n");
  }

  const lines: string[] = [
    "## Fleet Members",
    "",
    `You are part of the **${fleet.fleet.name}** fleet. Your peers:`,
    "",
  ];

  for (const peer of peers) {
    const peerSlack = slackChannel(peer);
    const userId = peerSlack?.bot_user_id ?? "TODO (run fleetmind slack discover)";
    const channels = (peerSlack?.channels ?? []).join(", ") || "(none configured)";
    const label = roleLabel(peer.role);
    lines.push(`- **${peer.name}** (${peer.emoji ?? ""}) — ${peer.description ?? ""}`);
    lines.push(`  - Slack user: \`${userId}\` (mention: \`<@${userId}>\`)`);
    lines.push(`  - Operates in: ${channels}`);
    lines.push(`  - Role: ${label}`);
    lines.push("");
  }

  lines.push("Use this roster when delegating, replying, or coordinating across channels.");
  lines.push("");

  return lines.join("\n");
}

function applyPlaceholders(text: string, agent: AgentConfig, fleet?: Fleet): string {
  let result = text
    .replaceAll("{{NAME}}", agent.name)
    .replaceAll("{{ID}}", agent.id)
    .replaceAll("{{EMOJI}}", agent.emoji ?? "")
    .replaceAll("{{DESCRIPTION}}", agent.description ?? "")
    .replaceAll("{{SOUL_BODY}}", agent.persona?.soul ?? "");

  if (fleet !== undefined) {
    result = result.replaceAll("{{FLEET_ROSTER}}", buildFleetRoster(fleet, agent));
  }

  return result;
}

/**
 * Absolute path to the fleetmind package root (where `openclaw/` lives).
 * Resolved from this file's own location so it works regardless of what
 * directory the operator runs `fleetmind` from.
 */
const PACKAGE_ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "../..");

/**
 * Shared workspace template dir (relative to the package root). Holds files
 * that are byte-identical across every bot type (HEARTBEAT.md, MEMORY.md,
 * TOOLS.md) so there is exactly one copy to edit instead of one per role.
 * Role-specific dirs (openclaw/<bot-type>/workspace/) still take priority —
 * this is only a fallback when the role dir doesn't ship the file itself.
 */
const SHARED_WORKSPACE_TEMPLATE_DIR = "openclaw/_shared/workspace";

function readRoleTemplate(role: string, filename: string): string | null {
  const dir = workspaceTemplatePath(role) ?? workspaceTemplatePath("worker")!;
  const filePath = path.resolve(PACKAGE_ROOT, dir, filename);
  if (fs.existsSync(filePath)) return fs.readFileSync(filePath, "utf8");

  const sharedPath = path.resolve(PACKAGE_ROOT, SHARED_WORKSPACE_TEMPLATE_DIR, filename);
  if (fs.existsSync(sharedPath)) return fs.readFileSync(sharedPath, "utf8");

  return null;
}

// =============================================================================
// Workspace file generators
// =============================================================================

function soulMd(agent: AgentConfig): string {
  return `# SOUL.md — ${agent.name}\n\n${agent.persona.soul}\n`;
}

function agentsMd(agent: AgentConfig): string {
  const role = agent.orchestrator
    ? "Orchestrator: coordinates the fleet and delegates to specialist agents."
    : "Specialist: handles delegated tasks from the orchestrator.";
  return `# AGENTS.md — ${agent.name} (${agent.emoji})\n\n${agent.description}\n\n## Role\n\n${role}\n\n## Memory\n\n- Daily notes: \`memory/YYYY-MM-DD.md\`\n- Long-term: \`MEMORY.md\`\n\n## Tools\n\nSkills define available tools. Check TOOLS.md for environment-specific notes.\n`;
}

function identityMd(agent: AgentConfig): string {
  const role = agent.orchestrator ? "Orchestrator" : "Specialist";
  return `# IDENTITY.md\n\n- **Name:** ${agent.name}\n- **Emoji:** ${agent.emoji}\n- **Role:** ${role}\n- **Description:** ${agent.description}\n`;
}

const USER_MD = `# USER.md — About Your Human\n\n_Populated during onboarding._\n\n- **Name:**\n- **Timezone:**\n- **Notes:**\n`;

function writeFile(filePath: string, content: string, dryRun: boolean): void {
  if (dryRun) return;
  fs.mkdirSync(path.dirname(filePath), { recursive: true });
  fs.writeFileSync(filePath, content, "utf-8");
}

// =============================================================================
// Public API
// =============================================================================

export async function provisionAgent(
  fleet: Fleet,
  agent: AgentConfig,
  dryRun: boolean,
  /** Local base directory for rendered output. Defaults to process.cwd().
   *  Workspace files are written to <localBase>/rendered/workspaces/<agent_id>/.
   *  workspace_base from fleet config is the EC2-side path; it is NOT used as a
   *  local mkdir target. */
  localBase: string = process.cwd()
): Promise<void> {
  // Local render target: ./rendered/workspaces/<agent_id>/ — consistent with
  // openclaw_json and terraform_vars which both go to ./rendered/.
  // workspace_base remains the EC2-side path (consumed by user-data and the
  // future deploy transport); it must NOT be used as a local mkdir target.
  const workspace = path.join(localBase, "rendered", "workspaces", agent.id);

  if (!dryRun) fs.mkdirSync(workspace, { recursive: true });

  const role = agent.role ?? "worker";

  const soulTemplate = readRoleTemplate(role, "SOUL.md");
  const soulContent = soulTemplate !== null
    ? applyPlaceholders(soulTemplate, agent, fleet)
    : soulMd(agent);
  writeFile(path.join(workspace, "SOUL.md"), soulContent, dryRun);

  const agentsTemplate = readRoleTemplate(role, "AGENTS.md");
  const agentsContent = agentsTemplate !== null
    ? applyPlaceholders(agentsTemplate, agent, fleet)
    : agentsMd(agent);
  writeFile(path.join(workspace, "AGENTS.md"), agentsContent, dryRun);

  const identityTemplate = readRoleTemplate(role, "IDENTITY.md");
  const identityContent = identityTemplate !== null
    ? applyPlaceholders(identityTemplate, agent, fleet)
    : identityMd(agent);
  writeFile(path.join(workspace, "IDENTITY.md"), identityContent, dryRun);

  // HEARTBEAT.md — operator-scaffolded, bots add task sections. Always ship
  // so pull-self's section merge can update the AUTO-tagged operator sections
  // while preserving bot-added task sections. Content is identical across every
  // role, so it lives once in SHARED_WORKSPACE_TEMPLATE_DIR and is picked up via
  // readRoleTemplate's fallback rather than duplicated per bot-type dir.
  const heartbeatTemplate = readRoleTemplate(role, "HEARTBEAT.md");
  if (heartbeatTemplate !== null) {
    writeFile(path.join(workspace, "HEARTBEAT.md"), heartbeatTemplate, dryRun);
  }

  // MEMORY.md — shipped on every push so the section merge can update
  // AUTO-tagged operator sections (e.g. Active Tasks) while preserving
  // everything the bot has written in untagged sections. Same shared-file
  // treatment as HEARTBEAT.md above.
  const memoryTemplate = readRoleTemplate(role, "MEMORY.md");
  if (memoryTemplate !== null) {
    writeFile(path.join(workspace, "MEMORY.md"), memoryTemplate, dryRun);
  }

  // TOOLS.md — only create if missing. pull-self protects this file from
  // being overwritten on existing agents, so this seeds it once for new agents.
  // Sourced from SHARED_WORKSPACE_TEMPLATE_DIR for every role (pm-bot and
  // backend-worker-bot previously shipped no TOOLS.md template at all, even
  // though every role's AGENTS.md boot sequence instructs the agent to read
  // TOOLS.md — the shared fallback closes that drift for all four bot types).
  const toolsMdTemplate = readRoleTemplate(role, "TOOLS.md");
  const toolsMdPath = path.join(workspace, "TOOLS.md");
  if (toolsMdTemplate !== null && !fs.existsSync(toolsMdPath)) {
    writeFile(toolsMdPath, applyPlaceholders(toolsMdTemplate, agent, fleet), dryRun);
  }

  // USER.md — only create if missing (don't overwrite customized versions)
  const userMdPath = path.join(workspace, "USER.md");
  if (!fs.existsSync(userMdPath)) {
    writeFile(userMdPath, USER_MD, dryRun);
  }

  // Install skills via three-tier resolver
  if (agent.skills.length > 0) {
    const skillsDir = path.join(workspace, "skills");
    if (!dryRun) fs.mkdirSync(skillsDir, { recursive: true });

    for (const skill of agent.skills) {
      await installSkill(skill, skillsDir, fleet, dryRun);
    }
  }

  log.dim(`  workspace → ${workspace}`);
}

export async function provisionFleet(
  fleet: Fleet,
  dryRun = false,
  /** Local base directory for rendered output. Defaults to process.cwd(). */
  localBase: string = process.cwd()
): Promise<void> {
  log.info(`\nFleetMind — provisioning fleet ${fleet.fleet.name}`);
  if (fleet.fleet.client) log.info(`  Client: ${fleet.fleet.client}`);
  log.info(`  Agents: ${fleet.agents.list.length}`);
  if (dryRun) log.warn("  Dry run — no changes will be made\n");

  for (const agent of fleet.agents.list) {
    log.step(`${agent.emoji} ${agent.name}...`);
    await provisionAgent(fleet, agent, dryRun, localBase);
    log.ok(`${agent.emoji} ${agent.name}`);
  }

  log.success("\nFleet provisioned.");
  log.info(`  Workspaces written to ${path.join(localBase, "rendered", "workspaces")}/`);
  log.info("  Next: run `fleetmind render` to generate openclaw.json");
}

export function diffFleet(
  fleet: Fleet,
  /** Local base directory for rendered output. Defaults to process.cwd(). */
  localBase: string = process.cwd()
): string[] {
  const changes: string[] = [];

  for (const agent of fleet.agents.list) {
    const workspace = path.join(localBase, "rendered", "workspaces", agent.id);

    if (!fs.existsSync(workspace)) {
      changes.push(`[+] Create workspace for ${agent.emoji} ${agent.name} at ${workspace}`);
      continue;
    }

    const soulPath = path.join(workspace, "SOUL.md");
    if (!fs.existsSync(soulPath)) {
      changes.push(`[+] ${agent.name}: create SOUL.md`);
    } else {
      const current = fs.readFileSync(soulPath, "utf-8").trim();
      const expected = soulMd(agent).trim();
      if (current !== expected) changes.push(`[~] ${agent.name}: update SOUL.md`);
    }

    const skillsDir = path.join(workspace, "skills");
    for (const skill of agent.skills) {
      const skillPath = path.join(skillsDir, skill.name);
      if (!fs.existsSync(skillPath)) {
        const source = skill.source ?? "client";
        const ver = skill.version ? `@${skill.version}` : "@latest";
        changes.push(`[+] ${agent.name}: install skill ${skill.name}${ver} (${source})`);
      }
    }

  }

  return changes.length > 0 ? changes : ["No changes detected."];
}
